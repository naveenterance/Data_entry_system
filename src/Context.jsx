import React, { createContext, useContext, useReducer } from "react";

const Context = createContext();

const ContextProvider = ({ children }) => {
  const initialState = {
    name: "",
    email: "",
    submittedData: [],
  };

  const reducer = (state, action) => {
    switch (action.type) {
      case "UPDATE_FIELD":
        return { ...state, [action.field]: action.value };
      case "RESET":
        return { ...state, name: "", email: "" };
      case "ADD":
        return {
          ...state,
          submittedData: [...state.submittedData, action.formData],
        };
      default:
        return state;
    }
  };

  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <Context.Provider value={{ state, dispatch }}>{children}</Context.Provider>
  );
};

export const Contexthook = () => {
  return useContext(Context);
};

export default ContextProvider;
