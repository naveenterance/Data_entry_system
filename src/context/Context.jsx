import React, { createContext, useContext, useReducer } from "react";

export const Context = createContext();

export const ContextProvider = ({ children }) => {
  const initialState = {
    firstName: "",
    lastName: "",
    email: "",
    phone: "",
    country: "",
    age: "",
    street: "",
    town: "",
    postcode: "",
    visits: "0",
    entries: [],
  };

  const reducer = (state, action) => {
    switch (action.type) {
      case "UPDATE_FIELD":
        return { ...state, [action.field]: action.value };
      case "RESET":
        return {
          ...state,
          firstName: "",
          lastName: "",
          email: "",
          phone: "",
          country: "",
          age: "",
          street: "",
          town: "",
          postcode: "",
          visits: "0",
        };
      case "ADD":
        return {
          ...state,
          entries: [...state.entries, action.formData],
        };
      default:
        return state;
    }
  };

  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <Context.Provider value={{ state, dispatch }}>{children}</Context.Provider>
  );
};

export const Contexthook = () => {
  return useContext(Context);
};
