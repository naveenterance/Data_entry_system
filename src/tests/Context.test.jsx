import React from "react";
import { render, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";

import { ContextProvider, Contexthook } from "../context/Context";

const MockChild = () => {
  const { state, dispatch } = Contexthook();

  const handleUpdateField = () => {
    dispatch({ type: "UPDATE_FIELD", field: "firstName", value: "John" });
    dispatch({ type: "UPDATE_FIELD", field: "lastName", value: "Brenann" });
    dispatch({ type: "UPDATE_FIELD", field: "email", value: "jb@gmail.com" });
    dispatch({ type: "UPDATE_FIELD", field: "phone", value: "12345678" });
  };

  return (
    <div>
      <span data-testid="firstName">{state.firstName}</span>
      <span data-testid="lastName">{state.lastName}</span>
      <span data-testid="email">{state.email}</span>
      <span data-testid="phone">{state.phone}</span>
      <button data-testid="submit" onClick={handleUpdateField}>
        Submit
      </button>
    </div>
  );
};

describe("Context", () => {
  it("renders without crashing ", () => {
    const { container } = render(
      <ContextProvider>
        <MockChild />
      </ContextProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it("updates context state when dispatching UPDATE_FIELD action", () => {
    const { getByTestId } = render(
      <ContextProvider>
        <MockChild />
      </ContextProvider>
    );

    expect(getByTestId("firstName")).toHaveTextContent("");
    expect(getByTestId("lastName")).toHaveTextContent("");
    expect(getByTestId("email")).toHaveTextContent("");
    expect(getByTestId("phone")).toHaveTextContent("");

    fireEvent.click(getByTestId("submit"));

    expect(getByTestId("firstName")).toHaveTextContent("John");
    expect(getByTestId("lastName")).toHaveTextContent("Brenann");
    expect(getByTestId("email")).toHaveTextContent("jb@gmail.com");
    expect(getByTestId("phone")).toHaveTextContent("12345678");
    const phone = getByTestId("phone").textContent;
    expect(typeof phone).toBe("string");
    expect(/^[0-9]+$/.test(phone)).toBe(true);
    const email = getByTestId("email").textContent;
    expect(typeof email).toBe("string");
    expect(/^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$/.test(email)).toBe(true);
  });
});
