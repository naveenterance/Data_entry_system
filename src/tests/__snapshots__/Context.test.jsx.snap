// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context renders without crashing  1`] = `
<div>
  <div>
    <span
      data-testid="firstName"
    />
    <span
      data-testid="lastName"
    />
    <span
      data-testid="email"
    />
    <span
      data-testid="phone"
    />
    <button
      data-testid="submit"
    >
      Submit
    </button>
  </div>
</div>
`;
